cvat/apps/dataset_manager/bindings.py

```diff
@@ -2151,10 +2151,64 @@ def reduce_fn(acc, v):
                 raise CvatImportError("Image {}: can't import annotation "
                     "#{} ({}): {}".format(item.id, idx, ann.type.name, e)) from e
 
-    for track in tracks.values():
-        track['elements'] = list(track['elements'].values())
-        instance_data.add_track(instance_data.Track(**track))
+    def append_necessary_outside_attribute(shapes):
+        new_shapes = []
+        prev_shape = None
+        # infer the keyframe shapes and keep only them
+        for shape in shapes:
+            cur_is_visible = shape is not None and not shape.outside
+            if not cur_is_visible:
+                prev_shape = shape
+                continue
+
+            prev_is_visible = prev_shape is not None and not prev_shape.outside
+            if prev_shape is None or shape.keyframe:
+                shape = shape._replace(keyframe=True)
+                new_shapes.append(shape)
+            elif not prev_is_visible or (
+                has_gap := prev_shape.frame + instance_data.frame_step < shape.frame
+            ):
+                if has_gap:
+                    prev_shape = prev_shape._replace(outside=True, keyframe=True,
+                        frame=prev_shape.frame + instance_data.frame_step)
+                    new_shapes.append(prev_shape)
 
+                shape = shape._replace(keyframe=True)
+                new_shapes.append(shape)
+
+            prev_shape = shape
+
+        if prev_shape and not prev_shape.outside and (
+            prev_shape.frame + instance_data.frame_step <= stop_frame
+            # has a gap before the current instance segment end
+        ):
+            prev_shape = prev_shape._replace(outside=True, keyframe=True,
+                frame=prev_shape.frame + instance_data.frame_step)
+            new_shapes.append(prev_shape)
+
+        return new_shapes
+
+    stop_frame = int(instance_data.meta[instance_data.META_FIELD]['stop_frame'])
+    for track_id, track in tracks.items():
+        track['shapes'].sort(key=lambda t: t.frame)
+        track['shapes'] = append_necessary_outside_attribute(track['shapes'])
+
+        if ann.type == dm.AnnotationType.skeleton:
+            new_elements = {}
+            for element_id, element in track['elements'].items():
+                element.shapes.sort(key=lambda t: t.frame)
+                new_element_shapes = append_necessary_outside_attribute(element.shapes)
+                new_elements[element_id] = instance_data.Track(
+                                label=element.label,
+                                group=element.group,
+                                source=element.source,
+                                shapes=new_element_shapes,
+                            )
+            track['elements'] = new_elements
+
+        if not(len(track['shapes']) == 0 or len(track['elements']) == 0):
```

Comment:
```suggestion
        if track['shapes']) or track['elements']:
```
We need to add only if there's anything to be added.